readme.md
# MGRS-Mapper (development repo)

A development repository for MGRS-Mapper application

---

## TODO

- [x] Next steps would be integrating the affiliation dropdown and adding a mutation observer to dynamically update the SVG in the panel
- [x] After that you need to add in the rest of the standard affiliations (hostile, templated, etc). There might be an issue with stroke-dasharray needing to be added to the matchPaths() function
- [x] Create a skeleton symbol object that contains all the decorator information for all affiliations
- [x] Add in the equipment icons (will require some more tweaking)
- [x] matchPaths() function and testObj need to be renamed. Come on man
- [x] The "demo-width-class" class in the html file needs to be changed to something less stupid
- [x] Code split your symbol object and affiliation object out into their own files
- [x] Modifier 1
- [x] Modifier 2
- [x] Unit Size modifier
- [x] "In Flight" outline modifiers for aviation symbols
- [x] Symbol text for Unique Designation and Higher formation
- [x] Task Force amplifier
- [x] Reinforced/Reduced modifier
- [x] Activity conversion (would require some logic, eg- a grenade launcher cannot be an activity)
- [x] Installation conversion (would require some logic, eg- a grenade launcher cannot be an installation)
- [x] Command Post amplifier
- [x] Tactical Mission Tasks
- [x] Graphic Control Measures
- [ ] Skeleton Screen preloader (Should be done right before launch)
- [x] Remove files and folders from gitignore in your repo [As seen here](http://www.codeblocq.com/2016/01/Untrack-files-already-added-to-git-repository-based-on-gitignore/)
- [ ] When typing slow in the symbol search field it will display "No Results Found". Find a way to only enable the search dropdown if the char length is at least 3
- [ ] Global vars need cleaning up. Use imports
- [ ] Helper info-window on the Mod1/2 dropdowns are needed. Most soldier's don't know this stuff
- [x] Create a folder for object JS files and rename them. Things are going to get more complicated as we add in Tactical Mission Tasks, Graphic Control Measures and Task Force Amps
- [ ] Mod1 helper text has word wrap. Fix it to 1 line
- [x] Fix the symbol panel so that amplifiers don't get clipped
- [x] Decide on using Bottom App Bar or Infowindow menu like on the Original pushbar site?
- [x] The select menus are too similar. Each should have an icon that differentiates them
- [x] Add surface menu
- [x] Configure CSS hover state for surface menu
- [ ] Add helper text to select boxes
- [x] Add SVG manipulation (drag-n-drop & resize handles)
- [x] Begin Leaflet maps implementation...
- [ ] Update repository key in package.json on production
- [x] Commonly used symbols pane
- [x] Logo in header
- [ ] Semantic HTML check. (eg- The "Dismissible Drawer" in the top app bar has a class of 'mdc-top-app-bar\_\_title')
- [ ] Sitemap file and Humans.txt
- [ ] SPA to actual website? Need to have multiple pages for this
- [ ] Update the "site updated" date in the humans.txt on production

## DEFAULT VALUES

### Path:

- d: REQUIRED
- fill: none
- stroke: black
- strokeWidth: 4

### Circle:

- cx: REQUIRED
- cy: REQUIRED
- r: REQUIRED
- fill: Black
- stroke: Black
- strokeWidth: 4

### Text:

- symbolText: REQUIRED
- x: REQUIRED
- y: REQUIRED
- text-anchor: REQUIRED
- font-size: REQUIRED
- fontFamily: Arial
- fontWeight: 30
- stroke: none
- strokeWidth: 4
- fill: black


changelog.md
# Change Log Notes

---

## 18MAR20

- Added 31 Mod2 symbols
- Finished Mod2 symbols
- Added 5 Command Post symbols
- Finished Command Post symbols
- Added 48 Tactical Mission Tasks
- Finished Tactical Mission Tasks symbols
- Added 144 Graphic Control Measures
- Finished Graphic Control Measures
- Refactoring militarySymbols.js
  - Starting lines: 18,891
  - Starting Size: 445.01KB
  - Ending lines: 10,986
  - Ending Size: 279.48KB
- Refactoring mod1.js
  - Starting lines: 3867
  - Starting Size: 90.8KB
  - Ending lines: 3718
  - Ending Size: 87.9KB

---

## 16MAR20

- Greatly optimized Resizer() function. Previously when the user would click on the symbolSelect menu, this function would resize ALL the symbols. This took nearly 700ms and there was a noticeable lag between clicking the menu and when it would appear. I added the functionality to check whether the symbol is visible on the screen and then only resize those symbols. When the user scrolls down it will continually check if the position of the menu and attempt to run the resize function when the scrollTop threshold is broken. This new functionality now takes 53ms to run.
- Added async/await to bounceInAnimation() and setSelectMenuTextContent() which greatly increased the speed of addSymbolsAndModsToList(). Bringing the time down from 450ms to 96ms
- Added PreloadWebpackPlugin to webpack.common.js. Max Potential First Input Delay lowered from ~2500ms to 230ms
  - On further inspection this does not always work. Might be an issue with running lighthouse on the dev build
- Added preconnect rel attribute on google fonts stylesheet
- Fixed bug preventing 100K grids from drawing on zoom levels less than 6. This was due to instantiating leaflet bounds in the initialize method
- Centered map on United States and brought the zoom level out so the 100K grids are not drawing on the map during page load
- Moved Roboto to head so I could add the preload attribute
- Fixed async bug where if the user searches for a symbol and clicks the delete button the app would crash
- Fixed async bug where if the user clears out a search field it would crash the app
- Fixed bug in Resizer function which caused FireFox to shit out.
- Fixed async bug where a user clicks a popular icon but there is text in the search field, so the app would explode.
- Fixed bug where the task force amplifier would cause the app the explode. Apparently MainMS was missing an echelon setting when I instantiated it
- Fixed bug where Aviation Equipment did NOT convert into flying outline
- Fixed bug where the symbol list would not show up if a user had any text inputted into the uniqueDesignation and higherFormation fields
- Fixed bbox bug on the 2nd element in the command post list
- Fixed bug where User Text control was outputting dark blue instead of light blue
- Fixed bug where Grid Overlay toggle for 100K labels were showing up as enabled despite being on zoom level 6. I have them disabled for that zoom level due to performance issues.
- Mod2 starting count: 20, ending: 30

---

## 14MAR20

- Added an async/await chain inside addSymbolsAndModsToList(). This has a great impact on reducing the page load times on the dev site

|             Metric              | Before | After  |
| :-----------------------------: | ------ | ------ |
|        Lighthouse Score         | 0      | 19     |
|     First Contentful Paint      | 45.2s  | 2.1s   |
|           Speed Index           | 45.2s  | 13.4s  |
|       Time to Interactive       | 46.9s  | 52.6s  |
|     First Meaningful Paint      | 45.7s  | 47.2s  |
|         First CPU Idle          | 45.7s  | 47.2s  |
| Max Potential First Input Delay | 2990ms | 2030ms |

- Had to call it quits early, got called over to Randy's for St Paddy's day celebration (despite the fucking coronavirus apocalypse).
- Since Web Workers don't interact with the DOM I can't use it with addSymbolsAndModsToList() but I could potentially load all those symbol object files?

---

## 11MAR20

- Disabled click propagation on symbols. Now when a user clicks on a symbol the map won't zoom in.
- Added a Leaflet control that allows the user to add text markers to the map
- Implemented the popup feature for text markers, so now they can be deleted!
- Added colored text selector for the User Text Input Leaflet Control
- Added the Leaflet.Draw plugin for more user interaction
- Ran into issues with getting a printing plugin for leaflet. I think I'll just roll my own. Leaflet.EasyPrint would work but when the print job is done all the markers would disappear. Look into [Print.js](https://printjs.crabbly.com/)
- Lighthouse Prod build:
  - Score: 8
  - First Contentful Paint: 11.4s
  - DOM Size: 7,501 elements
  - chunk.js: 569KB
  - app.js: 352KB
- Lighthouse Dev Build:
  - Score: 0
  - First Contentful Paint: 45.3s
  - DOM Size: 7,502 elements
  - chunk.js: 5.2MB
  - app.js: 2.7MB
- For some reason the most popular symbols event listener was duplicated verbatim... right below each other and I didn't catch it.
- Disabling the 100K grids on page load will bring the Dev build Lighthouse score from 0 to 20 and lowers the First Contentful Paint from 45 seconds to 1.7 seconds
- TODO: Combine Leaflet.DumbMGRS.js + map.js + moveSymbol.js
- TODO: Combine app.js + index.js
- I'm sick of these dependency cycle warnings. I have no idea how to fix them.

scratchpad.js
import { selectAffiliation } from './mdcComponents';

// When handleRemovePlayer is invoked we iterate through the players array in state and filter out only the player objects whose ID does NOT equal the ID passed into handleRemovePlayer
handleRemovePlayer = (id) => {
  this.setState(prevState => ({
    players: prevState.players.filter(player => player.id !== id),
  }));
};
// manifest.json provides metadata used when your web app is added to the
// homescreen on Android. See https://developers.google.com/web/fundamentals/engage-and-retain/web-app-manifest/


const bs = {
  text_1: {
    symbolText: 'ASLT',
    x: '100',
    y: '71',
    fontSize: '22',
    textAnchor: 'middle',
  },
};

const propertyToModify = {
  text_1: {
    fontSize: '19',
  },
};


selectMod1.menu_.items.map((key) => {
  const mod1Element = key.querySelectorAll('li figure svg g.outline path')[0];
  const mod1SVGContainer = mod1Element.parentElement.parentElement;
  mod1Element.setAttributeNS(null, 'fill', 'transparent');
  mod1Element.setAttributeNS(null, 'stroke-width', '0');
  mod1Element.setAttributeNS(null, 'd', '');
  mod1SVGContainer.setAttributeNS(null, 'viewBox', '80 55 40 20');
  mod1SVGContainer.style.transform = 'scale(0.75)';
});


selectSymbol.menu_.items.forEach((key) => {
  key.addEventListener('mouseenter', (event) => {
    console.log(event.target);
  });
});


const circ = document.querySelector('.newSVG > svg');
const miss = circ.querySelector('g.decorator');
miss.style.transformOrigin = '100px 100px'; // transform from center (cx, cy)
miss.style.transform = 'scale(0.75)';
const mod1 = circ.querySelector('g.mod1');
mod1.style.transform = `translateY(-${circ.viewBox.baseVal.x / circ.viewBox.baseVal.y * 21}px)`
// mod1.style.transform = 'translateY(-10%)';

  [selectSymbol, selectAffiliation, selectUnitSize, selectMod1].forEach((key) => {
    key.selectedText_.textContent = key.value;
    key.listen('MDCSelect:change', () => {
      new MilSym('.newSVG', selectSymbol.value, selectAffiliation.value, selectUnitSize.value, `${selectMod1.value || 'None'}`).placeSymbol();
    });
  });

// TODO: If you change the MilSym class to accept an object, then you can instantiate it like this: "new MilSym({location: '.test', symbol: 'Infantry', affiliation: 'friendly', echelon: 'team', mod1: 'Foraging'}).placeSymbol();". The only value I can see from this is making it easier to call this class since you can indent object keys on new lines.
// TODO: get mod1Data() and get mod2Data() in  MilSym are almost identical, there should be a way to combine them

class tttt {
  constructor() {
    this.reinforced = reinforcedSwitch.checked;
    this.reduced = reducedSwitch.checked;
    this.reinforcedAndReduced = this.reinforced && this.reduced;
  }
}

class RRSwitches {
  constructor() {
    this.reinforced = reinforcedSwitch.checked;
    this.reduced = reducedSwitch.checked;
    this.reinforcedAndReduced = this.reinforced && this.reduced;
    this.value = '';
    return this.checkSwitches(); //! You could use this on MilSym instead of using .placeSymbol() every time
  }

  checkSwitches() {
    switch (true) {
      case this.reinforcedAndReduced:
        console.log('I am reinforced and reduced!');
        reducedSwitch.root_.dataset.value = '±';
        reinforcedSwitch.root_.dataset.value = '';
        this.value = '±';
        break;
      case this.reinforced:
        console.log('I am reinforced');
        reinforcedSwitch.root_.dataset.value = '+';
        reducedSwitch.root_.dataset.value = '';
        this.value = '+';
        break;
      case this.reduced:
        console.log('I am Reduced!');
        reducedSwitch.root_.dataset.value = '–';
        reinforcedSwitch.root_.dataset.value = '';
        this.value = '–';
        break;
      default:
        console.log('I am nothing!');
        reducedSwitch.root_.dataset.value = '';
        reinforcedSwitch.root_.dataset.value = '';
        this.value = '';
        break;
    }
}
}

function checkSwitches() {
  this.reinforced = reinforcedSwitch.checked;
  this.reduced = reducedSwitch.checked;
  this.reinforcedAndReduced = this.reinforced && this.reduced;
  switch (true) {
    case this.reinforcedAndReduced:
      console.log('I am reinforced and reduced!');
      reducedSwitch.root_.dataset.value = '±';
      reinforcedSwitch.root_.dataset.value = '';
      break;
    case this.reinforced:
      console.log('I am reinforced');
      reinforcedSwitch.root_.dataset.value = '+';
      reducedSwitch.root_.dataset.value = '';
      break;
    case this.reduced:
      console.log('I am Reduced!');
      reducedSwitch.root_.dataset.value = '–';
      reinforcedSwitch.root_.dataset.value = '';
      break;
    default:
      console.log('I nothing!');
      reducedSwitch.root_.dataset.value = '';
      reinforcedSwitch.root_.dataset.value = '';
      break;
  }
}

const explicit = () => {
  const reinforced = reinforcedSwitch.checked;
  const reduced = reducedSwitch.checked;
  const reinforcedAndReduced = reinforced && reduced;
}

const rr = {
  get reinforced() {
    return reinforcedSwitch.checked;
  },
  get reduced() {
    return reducedSwitch.checked;
  },
  get reinforcedAndReduced() {
    return this.reinforced && this.reduced;
  },
  get truevals() {
    switch (true) {
      case this.reinforcedAndReduced:
        console.log('reinforcedAndReduced');
        reducedSwitch.root_.dataset.value = '±';
        reinforcedSwitch.root_.dataset.value = '';
        MainMS.reinforcedReduced = '±';
        MainMS.placeSymbol();
        break;
      case this.reinforced:
        console.log('reinforced');
        reinforcedSwitch.root_.dataset.value = '+';
        reducedSwitch.root_.dataset.value = '';
        MainMS.reinforcedReduced = '+';
        MainMS.placeSymbol();
        break;
      case this.reduced:
        console.log('reduced');
        reducedSwitch.root_.dataset.value = '–';
        reinforcedSwitch.root_.dataset.value = '';
        MainMS.reinforcedReduced = '–';
        MainMS.placeSymbol();
        break;
      default:
        console.log('none');
        reducedSwitch.root_.dataset.value = '';
        reinforcedSwitch.root_.dataset.value = '';
        MainMS.reinforcedReduced = '';
        MainMS.placeSymbol();
        break;
    }
  }
}

function gv(obj) {
  this.obj = obj;
  console.log(obj)
}

function name() {
  console.log(rr.truevals)
}

var rr6 = {
  reinforced: reinforcedSwitch.checked,
  reduced: reducedSwitch.checked,
  reinforcedAndReduced: reinforcedSwitch.checked && reducedSwitch.checked,
  get truevals() {




    if (this.reinforced) {
      console.log("reinforced true")
    }
    if (this.reduced) {
      console.log("reduced true")
    }
    if (this.reinforcedAndReduced) {
      console.log("reinforced & reduced true")
    }
  }
}

Object.entries(rrobj5).forEach(e => {
	e[1] ? console.log(e) : console.log("untrie")
})

async function lol(prop) {
  await this.prop = window.hasOwnProperty('MainMS');
  if (this.prop) {
    console.log("MainMS is in the window")
    setReinforcedReducedValues().truevals;
  }
}




[reducedSwitch, reinforcedSwitch].forEach(e => e.root_.dataset.value.length > 0 ? e.root_.dataset.value : null)

const m1 = document.querySelector('.mod1');
m1.style.transformBox = 'fill-box';
m1.style.transformOrigin = 'center center';
m1.classList.toggle('bounceIn');

////////

const HandleSelects = () => {
  console.log("I am Running");
  class SelectBoxes {
    constructor(sym, aff, size, mod1, mod2){
      this.sym = selectSymbol;
      this.aff = selectAffiliation;
      this.size = selectUnitSize;
      this.mod1 = selectMod1;
      this.mod2 = selectMod2;
      this.obj = {
          sym,
          aff,
          size,
          mod1,
          mod2,
        }
    }
  }
}
////

class HandleSelects {
  constructor() {
    this.sym = selectSymbol;
    this.aff = selectAffiliation;
    this.size = selectUnitSize;
    this.mod1 = selectMod1;
    this.mod2 = selectMod2;
    this.selectObjects = {
      selectSymbol,
      selectAffiliation,
      selectUnitSize,
      selectMod1,
      selectMod2,
    }
    return this.runFuncs();
  }

  get correctTextContent() {
    console.log("Running Correct Text Content");
    selectSymbol.selectedText_.textContent = selectSymbol.value;
    selectAffiliation.selectedText_.textContent = selectAffiliation.value.replace(/([A-Z])/g, ' / $1').replace(/^./, str => str.toUpperCase());
    selectUnitSize.selectedText_.textContent = selectUnitSize.value.replace(/([A-Z])/g, ' / $1').replace(/^./, str => str.toUpperCase());
    selectMod1.selectedText_.textContent = selectMod1.value;
    selectMod2.selectedText_.textContent = selectMod2.value;
  }

  get changeEvents() {
    console.log("Running Change Events")
    Object.values(this.selectObjects).forEach(key => {
        key.listen('MDCSelect:change', (event) => {
        switch (MainMS.type) {
          case 'Equipment':
            this.runFuncs();
            selectMod1.value !== 'None' ?  this.TransformModifiersOnEquipment('.newSVG > svg') : null;
            selectMod2.value !== 'None' ?  this.TransformModifiersOnEquipment('.newSVG > svg') : null;
            selectUnitSize.disabled = true;
            selectUnitSize.foundation_.adapter_.setSelectedIndex(0);
            uniqueDesignationField.disabled = true;
            uniqueDesignationField.value = '';
            deleteUniqueDesignationButton.root_.style.display = 'none';
            higherFormationField.disabled = true;
            higherFormationField.value = '';
            deleteHigherFormationButton.root_.style.display = 'none';
            reinforcedSwitch.disabled = true;
            reinforcedSwitch.checked = false;
            reducedSwitch.disabled = true;
            reducedSwitch.checked = false;
            return;

          case 'Land Unit':
            this.runFuncs();
            selectUnitSize.disabled = false;
            uniqueDesignationField.disabled = false;
            higherFormationField.disabled = false;
            reinforcedSwitch.disabled = false;
            reducedSwitch.disabled = false;

            if (event.target.classList.contains('symbol-select')) {
              // Only animate the symbol when a new symbol is clicked. This prevents the animation occurring on every single keyup in search field
              !searchField.input_.value ? document.querySelector('.newSVG > svg').setAttributeNS(null, 'class', 'animateSymbol') : null;
            }

            if (event.target.classList.contains('affiliation-select')) {
              selectSymbol.menu_.items.map((key2) => {
                new MilSym(`.symbolFigure[data-symbol-name="${key2.dataset.value}"]`, `${key2.dataset.value}`, `${selectAffiliation.value}`, 'none').placeSymbol();
              })
            }

            if (event.target.classList.contains('unit-size-select')) {
              this.BounceInAnimation('g.echelon');
            }

            if (event.target.classList.contains('mod1-select')) {
              this.BounceInAnimation('g.mod1');
            }

            if (event.target.classList.contains('mod2-select')) {
              this.BounceInAnimation('g.mod2');
            }
            break;
        }
      })

      key.listen('click', () => {
        switch (key.isMenuOpen_) {
          case selectSymbol.isMenuOpen_:
            this.Resizer('.symbolFigure svg');
            break;
          case selectUnitSize.isMenuOpen_:
            this.Resizer('.unitSizeFigure svg', 93, 33);
            break;
          case selectMod1.isMenuOpen_:
            this.Resizer('.mod1Figure svg');
            break;
          case selectMod2.isMenuOpen_:
            this.Resizer('.mod2Figure svg');
            break;
          default:
            break;
        }
      })
    })
  }

  TransformModifiersOnEquipment(equipmentOutline) {
     this.equipmentOutline = document.querySelector(equipmentOutline); // The equipment SVG you want to readjust
     this.equipmentDecorator = this.equipmentOutline.querySelector('g.decorator');
     this.mod1 = this.equipmentOutline.querySelector('g.mod1');
     this.mod2 = this.equipmentOutline.querySelector('g.mod2');
     this.equipmentDecorator.style.transformOrigin = '100px 100px'; // transform from center of circle (cx, cy)
     this.equipmentDecorator.style.transform = 'translateY(2%) scale(0.75)';
     // mod1.style.transform = `translateY(-${equipmentOutline.viewBox.baseVal.x / equipmentOutline.viewBox.baseVal.y * 21}px)`;
     this.mod1.style.transformOrigin = '100px 100px';
     this.mod1.style.transform = 'translateY(-11%) scale(0.85)';
     this.mod2.style.transformOrigin = '100px 140px';
     this.mod2.style.transform = 'scale(0.85)';
  }

  Resizer(symbolElement, width = 93, height = 64) {
    this.symbolElement = document.querySelectorAll(symbolElement);
    this.width = width;
    this.height = height;
    this.symbolElement.forEach((key) => {
      key.setAttributeNS(null, 'preserveAspectRatio', 'xMidYMid'); // this is a default value I believe
      key.setAttributeNS(null, 'viewBox', `${key.getBBox().x - 4} ${key.getBBox().y - 4} ${key.getBBox().width + 8} ${key.getBBox().height + 8}`);
      key.setAttributeNS(null, 'width', `${this.width}`);
      key.setAttributeNS(null, 'height', `${this.height}`);
    });
  }

  BounceInAnimation(location) {
    const bounceIn = document.querySelector(location);
    bounceIn.style.transformBox = 'fill-box';
    bounceIn.style.transformOrigin = 'center center';
    bounceIn.classList.toggle('bounceIn');
  }

  runFuncs() {
    this.correctTextContent;
    this.changeEvents;
    const MainMS = new MilSym('.newSVG',
      selectSymbol.value,
      selectAffiliation.value,
      selectUnitSize.value,
      `${selectMod1.value || 'None'}`,
      `${selectMod2.value || 'None'}`,
      uniqueDesignationField.value,
      higherFormationField.value,
      ''
      );
    window.MainMS = MainMS; //! Put this class instance in the global scope so it can be referenced and edited
    MainMS.placeSymbol();
  }
}


///////! no
HandleInputs(selectSymbol, selectAffiliation, selectUnitSize, selectMod1, selectMod2, uniqueDesignationField, higherFormationField, reinforcedSwitch, reducedSwitch)

const HandleInputs = (symbol, affiliation, unitSize, mod1, mod2, uniqueUnit, higherFormation, reinforced, reduced) => {
  this.symbol = symbol;
  this.affiliation = affiliation;
  this.unitSize = unitSize;
  this.mod1 = mod1;
  this.mod2 = mod2;
  this.uniqueUnit = uniqueUnit;
  this.higherFormation = higherFormation;
  this.reinforced = reinforced;
  this.reduced = reduced;

  const inputObjects = {
    symbol: {
      text: symbol.value,
      type: 'select',
      data: symbol,
    },
    affiliation,
  }
  return inputObjects
}

//! NEW SHIT

var HandleInputs = () => {
  const selectObjects = {
    selectSymbol: {
      text: selectSymbol.value,
      type: 'select',
      data: selectSymbol,
    },
    selectAffiliation: {
      text: selectAffiliation.value.replace(/([A-Z])/g, ' / $1').replace(/^./, str => str.toUpperCase()),
      type: 'select',
      data: selectAffiliation,
    },
    selectUnitSize: {
      text: selectUnitSize.value.replace(/([A-Z])/g, ' / $1').replace(/^./, str => str.toUpperCase()),
      type: 'select',
      data: selectUnitSize,
    },
    selectMod1: {
      text: selectMod1.value,
      type: 'select',
      data: selectMod1,
    },
    selectMod2: {
      text: selectMod2.value,
      type: 'select',
      data: selectMod2,
    },
  }

  const textFieldObjects = {
    uniqueDesignationField: {
      text: uniqueDesignationField.value,
      type: 'text',
      data: uniqueDesignationField,
    },
    higherFormationField: {
      text: higherFormationField.value,
      type: 'text',
      data: higherFormationField,
    },
  }

  const switchObjects = {
    reinforcedSwitch: {
      text: null,
      type: 'switch',
      data: reinforcedSwitch,
    },
    reducedSwitch: {
      text: null,
      type: 'switch',
      data: reducedSwitch,
    },
  }


  const runFuncs = () => {
    correctTextContent();
    const MainMS = new MilSym('.newSVG', selectSymbol.value, selectAffiliation.value, selectUnitSize.value, `${selectMod1.value || 'None'}`, `${selectMod2.value || 'None'}`, uniqueDesignationField.value, higherFormationField.value, ''); //! Remember to put reinforcedReducedValue() function back in there!!
    window.MainMS = MainMS; //! Put this class instance in the global scope so it can be referenced and edited
    MainMS.placeSymbol();
    //changeEvents();
  }

  const correctTextContent = () => {
    console.log("Running correctTextContent")
    Object.values(selectObjects).forEach(key => {
      key.data.selectedText_.textContent = key.text;
    })
  }

  // const changeEvents = () => {
  //   Object.values(selectObjects).forEach(key => {
  //     key.data.listen('MDCSelect:change', (event) => {
  //       switch (MainMS.type) {
  //         case 'Equipment':
  //           console.log("Equipment Found")
  //           return;
  //         case 'Land Unit':
  //           console.log("LAND UNIT")
  //           return;
  //         default:
  //           break;
  //       }
  //     })
  //   })
  // }
  Object.values(selectObjects).forEach(key => {
    key.data.listen('MDCSelect:change', (event) => {
      switch (MainMS.type) {
        case 'Equipment':
          console.log("Equipment Found")
          runFuncs();
          return;
        case 'Land Unit':
          console.log("LAND UNIT")
          runFuncs()
          return;
        default:
          break;
      }
    })
  })

  return runFuncs()
}
//Friendly Flight Outline: M 155,150 C 155,50 115,30 100,30 85,30 45,50 45,150
affiliationOutlineObject + selectAffiliation.value;

function prop() {
  const propertyToModify = {d: 'M 155,150 C 155,50 115,30 100,30 85,30 45,50 45,150',};
  const modifiedTarget = Object.assign({}, affiliationOutlineObject[selectAffiliation.value].d, propertyToModify);
  return modifiedTarget.d;
}


//Good example of using promises in the selectAffiliation dropdown
const changeCommandPostOutlines = new Promise((resolve, reject) => {
  console.log('resolving the commandpost');
  resolve(selectCommandPost.menu_.items.forEach(e => e.remove()));
  reject(new Error('In 10% of the cases, I fail. Miserably.'));
});

changeCommandPostOutlines.then(() => {
  addSymbolsAndModsToList(commandPostObject, 'commandpost', selectCommandPost);
  // Remove the outline of the default/none symbol
  document.querySelectorAll('.commandpostFigure svg g.outline')[0].remove();
}, (error) => {
  console.log(error);
});

// Bounce in Animation with Equipment logic.. Does not work that well
async function bounceInAnimation(location) {
  const bounceIn = document.querySelector(location);
  if (await MainMS.type === 'Equipment') {
    bounceIn.style.transform = 'translateY(0%) scale(0.85)';
    bounceIn.style.transformBox = 'border-box';
    bounceIn.style.transformOrigin = '100px 0px';
    bounceIn.classList.toggle('bounceIn');
  } else {
    // transformBox is crucial. Without this Mod1 will not scale from the center
    bounceIn.style.transformBox = 'fill-box';
    bounceIn.style.transformOrigin = 'center center';
    bounceIn.classList.toggle('bounceIn');
  }
}

// ! Async functions example
(async () => {
  // Don't listen to VS Code, await IS necessary for this
  await outlineGroup.append(outlineTemplated);
  outlineGroup.prepend(outline);
})().catch((err) => {
  console.error(err);
  setTimeout(() => {
    outlineGroup.append(outlineTemplated);
    outlineGroup.prepend(outline);
  }, 30);
});
// This async function also works
(async () => {
  await outlineGroup.append(outline, outlineTemplated);
  outline.insertAdjacentElement('afterend', outlineTemplated);
})();